Update "Dropping older Python versions" guide (#1471)
pypa/packaging.python.org: source/guides/dropping-older-python-versions.rst

Dropping support for older Python versions
==========================================

> Dropping support for older Python versions is supported by the standard :ref:`core-metadata` 1.2 specification via a :ref:`"Requires-Python" <core-metadata-requires-python>` attribute.

Reviewer: > support ... is supported

Can we change one of these words to something else to improve the writing style?

Reviewer: Also, perhaps using `s/a/the/` would be more grammatically correct.

Metadata 1.2+ clients, such as Pip 9.0+, will adhere to this specification by matching the current Python runtime and comparing it with the required version
in the package metadata. If they do not match, it will attempt to install the last package distribution that supported that Python runtime.

This mechanism can be used to drop support for older Python versions, by amending the ``Requires-Python`` attribute in the package metadata.

Requirements
------------

> This workflow requires that the user installing the package has at least Pip 9.0, or a client that supports the Metadata 1.2 specification.

Reviewer 2: This is fine as it is, but I'd suggest to just drop this sentence, pip 9.0 is 7 years old at this point. (As written on another review, my guess is that many commands in the PUG will not work with pip versions that are that old anyway.)

Author: > This is fine as it is

Does that mean you suggest restoring the list? I took out the `setuptools` requirement since this guide now focuses on `pyproject.toml`, and the `twine` requirement is already set in the linked distributing guide (but I could as well reinstate that).

Reviewer: A lot of LTS distro users (e.g. RHEL) may still be constrained to quite old versions. Here's what RHEL 8.3 has:
```console
$ podman run --rm -it -v "$(pwd):/io" ubi8/ubi:8.3 /usr/libexec/platform-python -m ensurepip --version    
pip 9.0.3
```

RHEL 7.9 has one instance of pip 8:
```console
[root@92c71f44a6f9 /]# yum info python27-python-pip
Loaded plugins: ovl, product-id, search-disabled-repos, subscription-manager

This system is not receiving updates. You can use subscription-manager on the host to register and assign subscriptions.

Installed Packages
Name        : python27-python-pip
Arch        : noarch
Version     : 8.1.2
Release     : 7.el7
Size        : 6.6 M
Repo        : installed
From repo   : ubi-server-rhscl-7-rpms
Summary     : A tool for installing and managing Python packages
URL         : http://www.pip-installer.org
License     : MIT and Python and ASL 2.0 and BSD and ISC and LGPLv2 and (ASL 2.0 or BSD)
Description : Pip is a replacement for `easy_install
            : <http://peak.telecommunity.com/DevCenter/EasyInstall>`_.  It uses mostly the
            : same techniques for finding packages, so packages that were made
            : easy_installable should be pip-installable as well.
```

Of course, it's quite old but a version hint is still relevant. `manylinux1` images that are still in use, are based on an even older CentOS 6, for example.

Reviewer 2: > Does that mean you suggest restoring the list?

Sorry, I meant that your modified version is fine as it is.

Maybe the version hint should stay but in a footnote?

> Dealing with the universal wheels

Reviewer: @chrysle this paragraph is specifically talking about `setuptools`, `bdist_wheel` as a `setuptools`' command that invokes the `wheel` project APIs internally and a CLI option (in the tip block below) that is supposed to be passed to the direct CLI invocation of `setup.py`, which is also discouraged.

So while you're on it, could to add clarifications that it's solely about the setuptools' mechanisms and maybe drop the suggestions to use the CLI flag?

   [bdist_wheel]
   universal = 0  # Make the generated wheels have "py3" tag

.. hint::

   Regarding :ref:`deprecated <setup-py-deprecated>` direct ``setup.py`` invocations,
   passing the ``--universal`` flag on the command line could override this setting.

Defining the Python version required
------------------------------------

1. Install twine
~~~~~~~~~~~~~~~~

Ensure that you have twine available at its latest version.
Steps:

.. tab:: Unix/macOS

    .. code-block:: bash

        python3 -m pip install --upgrade twine

.. tab:: Windows

    .. code-block:: bat

        py -m pip install --upgrade twine

2. Specify the version ranges for supported Python distributions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can specify version ranges and exclusion rules (complying with the :ref:`version-specifiers` specification),
such as at least Python 3. Or, Python 2.7, 3.4 and beyond:

.. code-block:: text

    Requires-Python: ">= 3"

>     Requires-Python: ">= 2.7, != 3.0.*, != 3.1.*, != 3.2.*, != 3.3.*"

Reviewer 3: Should this example be updated to use newer Python versions? I don't think most people reading this guide would be interested in supported Python 2.7 or 3.4-3.7.

Author: Done!

Those values can be set within your :file:`pyproject.toml`. The :ref:`requires-python` configuration field
corresponds to the ``Requires-Python`` metadata field.

.. code-block:: toml

   [build-system]
   ...

   [project]
   requires-python = ">= 3.8" # At least Python 3.8

For :ref:`setuptools` users, another way to achieve this is using the ``python_requires`` parameter
in your :file:`setup.cfg` config or the :file:`setup.py` script. ``setuptools < 61`` does not support
declaring the package metadata in :file:`pyproject.toml`.

Consult the ``setuptools`` `dependency-management`_ documentation for information about the appropriate
way to configure each of these files.

.. caution::

>         It is warned against adding upper bounds to the version ranges, e. g. ``">= 3.8 < 3.10"``. This can cause different errors

Reviewer 4: Use imperative voice, fix the missing comma, and avoid ambiguous "this":

```suggestion
        Avoid adding upper bounds to the version ranges, e. g. ``">= 3.8, < 3.10"``. Doing so can cause different errors
```

Defining the Python version required
------------------------------------

1. Install twine
~~~~~~~~~~~~~~~~

Ensure that you have twine available at its latest version.
Steps:

.. tab:: Unix/macOS

    .. code-block:: bash

        python3 -m pip install --upgrade twine

.. tab:: Windows

    .. code-block:: bat

        py -m pip install --upgrade twine

2. Specify the version ranges for supported Python distributions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can specify version ranges and exclusion rules (complying with the :ref:`version-specifiers` specification),
such as at least Python 3. Or, Python 2.7, 3.4 and beyond:

.. code-block:: text

    Requires-Python: ">= 3"
    Requires-Python: ">= 2.7, != 3.0.*, != 3.1.*, != 3.2.*, != 3.3.*"

> The way to set those values is within your :file:`pyproject.toml`. The :ref:`requires-python` field of the

Reviewer 2: I think “the way” is not really accurate because it's not in, e.g., Poetry. Maybe just `` :ref:`requires-python` `` to link to the pyproject.toml guide? (There's a note at the top of that guide, explaining that not all build backends use the `[project]` table.)

Reviewer: I don't think that Poetry should be special-cased as we're building guides around the accepted and widely implemented standards and not focus on tools that don't want to play ball. Besides, whenever they get to implementing it, it'll work for them too.

OTOH, I'd not focus on `pyproject.toml` solely as I still prefer a tool-specific `setup.cfg` myself. So the argument of not presenting it as the only possible way of declaring the metadata makes sense to me. I'd probably say something like along the lines of "the way declared by the standard/PEP X is this, but tool specific alternatives exist".

Reviewer 2: Yes, sorry for not being clear, I'm not suggesting to list how every tool does it, just not to lead the reader to believe that `[project]` is always used. This is already explained in a note near the beginning of the [pyproject.toml guide](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/), hence the proposal to link that document here.

``[project]`` table corresponds to the ``Requires-Python`` metadata field.

.. code-block:: toml

   [build-system]
   ...

   [project]
   requires-python = ">= 3.8" # At least Python 3.8

> For :ref:`setuptools` users, another way to achieve this is using the ``python_requires`` parameter
> in the call to ``setup`` within your :file:`setup.py` script.

Reviewer: Let's make this a bit more accurate:
```suggestion
For :ref:`setuptools` users, another way to achieve this is using the ``python_requires`` parameter in your :file:`setup.cfg` config or the :file:`setup.py` script.
``setuptools < 61`` does not support declaring the package metadata in :file:`pyproject.toml`.
```

> .. code-block:: python

Reviewer: Can we demonstrate using a declarative config instead of a script here?
```suggestion
.. code-block:: ini

   # setup.cfg

   [options]
   python_requires = >= 3.8
```

Reviewer: Also, `setuptools` now has comprehensive examples in the docs. See:
- https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#python-requirement
- https://setuptools.pypa.io/en/latest/userguide/declarative_config.html
- https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html

So perhaps linking them instead of having a tool-specific example spelled out, would be a good option.

    setup(
        # Your setup arguments
        python_requires='>= 3.8',
    )

> It is warned against adding upper bounds to the version ranges, e. g. ``">= 3.8 < 3.10"``. This can cause different errors

Reviewer: Could you add an admonition here? Like `.. caution::`? https://pradyunsg.me/furo/reference/admonitions/#caution

> and version conflicts. See the `discourse discussion`_ for more information.

Reviewer 2: Nit: reST labels normally use hyphens.

Reviewer: +1 to using dashes. Using snake case in contexts differing from Python code helps notices said difference visually. This is the reason why I use dashes in YAML identifiers and other places where it's allowed — it clearly communicates that this syntax is from some other context.

3. Validating the Metadata before publishing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Within a Python source package (the zip or the tar-gz file you download) is a text file called PKG-INFO.

> This file is generated by the build backend when it generates the source package.

Reviewer: Could you link the build backend term?

The file contains a set of keys and values, the list of keys is part of the PyPA standard metadata format.

You can see the contents of the generated file like this:

.. code-block:: bash

>     tar xf dist/my-package-1.0.0.tar.gz my-package-1.0.0/PKG-INFO -O

Reviewer: Why?

Validate that the following is in place, before publishing the package:

- If you have upgraded correctly, the Metadata-Version value should be 1.2 or higher.
- The ``Requires-Python`` field is set and matches your specification in the configuration file.

4. Publishing the package
~~~~~~~~~~~~~~~~~~~~~~~~~

Proceed as suggested in :ref:`Uploading your Project to PyPI`.

Dropping a Python version
-------------------------

In principle, at least metadata support for Python versions should be kept as long as possible, because
once that has been dropped, people still depending on a version will be forced to downgrade.
If however supporting a specific version becomes a blocker for a new feature or other issues occur, the metadata
``Requires-Python`` should be amended. Of course this also depends on whether the project needs to be stable and
well-covered for a wider range of users.

> Each version compatibility change should have an own release.

Reviewer: ```suggestion
Each version compatibility change should have its own release.
```

For example, you published version 1.0.0 of your package with ``Requires-Python: ">= 2.7"`` metadata.

> If you then update the version string to ``">= 3.5"``, and publish a new version 2.0.0 of your package, any users running Pip 9.0+ from version 2.7 will have version 1.0.0 of the package installed, and any ``>= 3.5`` users will receive version 2.0.0.

Reviewer: Not sure I agree with this at all. It's absolutely orthogonal to SemVer. The dependency resolvers will always select a runtime-compatible version for as long as this information is present in the package metadata.

Dropping a runtime support is not a feature change. And in SemVer, the major version release is for breaking APIs a project provides to the end-users. This metadata change may happen anytime and is a subject to different considerations that are dependent on the conventions of specific projects.

Let's not overload this guide with assumptions based on one specific schema for communicating API changes, that is known to be problematic anyway.

The maintainers can choose to synchronize dropping a runtime with a versioning scheme of their choice but that's not something that is mandatory for changing a metadata field value.

I think @jaraco mentioned this too somewhere.

Reviewer 4: Agreed on not making assumptions about versions schemes.

It's been my instinct that dropping a platform _isn't_ a breaking change because of the reasons described (it doesn't break anyone who's likely to download it).

> It may be a good idea to create a minor release, stating that it is the last one compatible with the Python version to be removed, just before dropping that version.

Reviewer: I don't think this advice makes sense anymore. It could be a patch release too. The idea here was that before `Requires-Python` became a thing, many projects were released without this bit of metadata. And this would cause the dependency resolver to treat the old releases as supporting any Python runtime. So if people were to add a high minimum Python version right away, the resolver would backtrack to the one immediately preceding release.

So this is a good advice for old project that never had this metadata field set. Once they start using it, they can update it anytime, regardless of any other pieces of the metadata.

> When dropping a Python version, it might also be rewarding to upgrade the project's code syntax generally, apart from updating the versions used in visible places (like the testing environment). Tools like pyupgrade_ can simplify this task.

Reviewer: Let's put this into an admonition like `.. hint::` or `.. tip::`.

Validate that the following is in place, before publishing the package:

- If you have upgraded correctly, the Metadata-Version value should be 1.2 or higher.
- The ``Requires-Python`` field is set and matches your specification in the configuration file.

4. Publishing the package
~~~~~~~~~~~~~~~~~~~~~~~~~

Proceed as suggested in :ref:`Uploading your Project to PyPI`.

Dropping a Python version
-------------------------

In principle, at least metadata support for Python versions should be kept as long as possible, because
once that has been dropped, people still depending on a version will be forced to downgrade.
If however supporting a specific version becomes a blocker for a new feature or other issues occur, the metadata
``Requires-Python`` should be amended. Of course this also depends on whether the project needs to be stable and
well-covered for a wider range of users.

Each version compatibility change should have its own release.

.. tip::

>         When dropping a Python version, it might also be rewarding to upgrade the project's code syntax generally, apart from updating the versions used in visible         places (like the testing environment). Tools like pyupgrade_ can simplify this task.

Reviewer 4: ```suggestion
        When dropping a Python version, it might also be rewarding to upgrade the project's code syntax generally, apart from updating the versions used in visible places (like the testing environment). Tools like pyupgrade_ or `ruff <https://docs.astral.sh/ruff/linter/>`_ can automate some of this work.
```
